velox/exec/tests/utils/TpchQueryBuilder.cpp

```diff
@@ -892,6 +894,90 @@ TpchPlan TpchQueryBuilder::getQ14Plan() const {
   return context;
 }
 
+TpchPlan TpchQueryBuilder::getQ15Plan() const {
+  std::vector<std::string> lineitemColumns = {
+      "l_suppkey", "l_shipdate", "l_extendedprice", "l_discount"};
+  std::vector<std::string> supplierColumns = {
+      "s_suppkey", "s_name", "s_address", "s_phone"};
+
+  const auto lineitemSelectedRowType = getRowType(kLineitem, lineitemColumns);
+  const auto& lineitemFileColumns = getFileColumnNames(kLineitem);
+  const auto supplierSelectedRowType = getRowType(kSupplier, supplierColumns);
+  const auto& supplierFileColumns = getFileColumnNames(kSupplier);
+
+  const std::string shipDateFilter = formatDateFilter(
+      "l_shipdate", lineitemSelectedRowType, "'1996-01-01'", "'1996-03-31'");
+
+  auto planNodeIdGenerator = std::make_shared<PlanNodeIdGenerator>();
+  core::PlanNodeId lineitemScanNodeIdSubQuery;
+  core::PlanNodeId lineitemScanNodeId;
+  core::PlanNodeId supplierScanNodeId;
+
+  auto revenue0 =
+      PlanBuilder(planNodeIdGenerator)
+          .tableScan(
+              kLineitem,
+              lineitemSelectedRowType,
+              lineitemFileColumns,
+              {shipDateFilter})
+          .capturePlanNodeId(lineitemScanNodeIdSubQuery)
+          .project(
+              {"l_suppkey as supplier_no",
+               "l_extendedprice * (1.0 - l_discount) as part_revenue"})
+          .partialAggregation(
+              {"supplier_no"}, {"sum(part_revenue) as total_revenue"})
+          .localPartition({})
+          .finalAggregation()
+          .planNode();
+
+  auto lineitemMaxRevenue =
+      PlanBuilder(planNodeIdGenerator)
+          .tableScan(
+              kLineitem,
+              lineitemSelectedRowType,
+              lineitemFileColumns,
+              {shipDateFilter})
+          .capturePlanNodeId(lineitemScanNodeId)
+          .project(
+              {"l_suppkey",
+               "l_extendedprice * (1.0 - l_discount) as part_revenue"})
+          .partialAggregation(
+              {"l_suppkey"}, {"sum(part_revenue) as total_revenue"})
+          .localPartition({})
+          .finalAggregation()
+          .partialAggregation({}, {"max(total_revenue) as max_revenue"})
+          .localPartition({})
+          .finalAggregation()
+          .planNode();
+
+  auto plan =
+      PlanBuilder(planNodeIdGenerator)
+          .tableScan(kSupplier, supplierSelectedRowType, supplierFileColumns)
+          .capturePlanNodeId(supplierScanNodeId)
+          .hashJoin(
+              {"s_suppkey"},
+              {"supplier_no"},
+              revenue0,
+              "",
+              {"s_suppkey", "s_name", "s_address", "s_phone", "total_revenue"})
+          .hashJoin(
```

Comment:
Would you share the output of printPlanWithStats? I wonder what are the join cardinalities and whether it makes sense to swap the join order.

Reply:
This is the output of Velox plan with stats, the other join order seems to be better:
Execution time: 7.71s
Splits total: 230, finished: 230
-- OrderBy[s_suppkey ASC NULLS LAST] -> s_suppkey:BIGINT, s_name:VARCHAR, s_address:VARCHAR, s_phone:VARCHAR, total_revenue:DOUBLE
   Output: 1 rows (143.94KB, 1 batches), Cpu time: 75.00us, Blocked wall time: 0ns, Peak memory: 1.00MB, Memory allocations: 12, Threads: 1
  -- HashJoin[INNER total_revenue=max_revenue] -> s_suppkey:BIGINT, s_name:VARCHAR, s_address:VARCHAR, s_phone:VARCHAR, total_revenue:DOUBLE
     Output: 1 rows (189B, 1 batches), Cpu time: 2.46ms, Blocked wall time: 7.66s, Peak memory: 2.00MB, Memory allocations: 8
     HashBuild: Input: 1 rows (32B, 1 batches), Output: 0 rows (0B, 0 batches), Cpu time: 71.00us, Blocked wall time: 0ns, Peak memory: 1.00MB, Memory allocations: 2, Threads: 1
        distinctKey0       sum: 1, count: 1, min: 1, max: 1
        queuedWallNanos    sum: 90.00us, count: 1, min: 90.00us, max: 90.00us
     HashProbe: Input: 100000 rows (14.24MB, 104 batches), Output: 1 rows (189B, 1 batches), Cpu time: 2.38ms, Blocked wall time: 7.66s, Peak memory: 1.00MB, Memory allocations: 6, Threads: 1
        queuedWallNanos    sum: 60.00us, count: 1, min: 60.00us, max: 60.00us
    -- HashJoin[INNER s_suppkey=supplier_no] -> s_suppkey:BIGINT, s_name:VARCHAR, s_address:VARCHAR, s_phone:VARCHAR, total_revenue:DOUBLE
       Output: 100000 rows (14.24MB, 104 batches), Cpu time: 24.90ms, Blocked wall time: 11.71ms, Peak memory: 5.00MB, Memory allocations: 162
       HashBuild: Input: 100000 rows (2.29MB, 98 batches), Output: 0 rows (0B, 0 batches), Cpu time: 11.01ms, Blocked wall time: 0ns, Peak memory: 4.00MB, Memory allocations: 52, Threads: 1
          distinctKey0       sum: 100001, count: 1, min: 100001, max: 100001
          queuedWallNanos    sum: 87.00us, count: 1, min: 87.00us, max: 87.00us
          rangeKey0          sum: 100001, count: 1, min: 100001, max: 100001
       HashProbe: Input: 100000 rows (974.87KB, 12 batches), Output: 100000 rows (14.24MB, 104 batches), Cpu time: 13.89ms, Blocked wall time: 11.71ms, Peak memory: 1.00MB, Memory allocations: 110, Threads: 1
          dataSourceLazyWallNanos    sum: 10.14ms, count: 36, min: 24.00us, max: 710.00us
          dynamicFiltersProduced     sum: 1, count: 1, min: 1, max: 1
          queuedWallNanos            sum: 71.00us, count: 1, min: 71.00us, max: 71.00us
      -- TableScan[table: supplier] -> s_suppkey:BIGINT, s_name:VARCHAR, s_address:VARCHAR, s_phone:VARCHAR
         Input: 100000 rows (974.88KB, 0 batches), Output: 100000 rows (974.87KB, 12 batches), Cpu time: 20.51ms, Blocked wall time: 0ns, Peak memory: 5.00MB, Memory allocations: 514, Threads: 1, Splits: 70
            dataSourceWallNanos       sum: 11.88ms, count: 82, min: 3.00us, max: 1.25ms
            dynamicFiltersAccepted    sum: 1, count: 1, min: 1, max: 1
            localReadBytes            sum: 0B, count: 1, min: 0B, max: 0B
            numLocalRead              sum: 0, count: 1, min: 0, max: 0
            numPrefetch               sum: 0, count: 1, min: 0, max: 0
            numRamRead                sum: 0, count: 1, min: 0, max: 0
            numStorageRead            sum: 0, count: 1, min: 0, max: 0
            prefetchBytes             sum: 0B, count: 1, min: 0B, max: 0B
            ramReadBytes              sum: 0B, count: 1, min: 0B, max: 0B
            skippedSplitBytes         sum: 0B, count: 1, min: 0B, max: 0B
            skippedSplits             sum: 0, count: 1, min: 0, max: 0
            skippedStrides            sum: 0, count: 1, min: 0, max: 0
            storageReadBytes          sum: 0B, count: 1, min: 0B, max: 0B
      -- Aggregation[FINAL [supplier_no] total_revenue := sum("total_revenue")] -> supplier_no:BIGINT, total_revenue:DOUBLE
         Output: 100000 rows (2.29MB, 98 batches), Cpu time: 7.49ms, Blocked wall time: 0ns, Peak memory: 4.00MB, Memory allocations: 140, Threads: 1
        -- LocalPartition[GATHER] -> supplier_no:BIGINT, total_revenue:DOUBLE
           Output: 200000 rows (4.59MB, 98 batches), Cpu time: 945.00us, Blocked wall time: 7.65s, Peak memory: 0B, Memory allocations: 0
           LocalPartition: Input: 100000 rows (2.29MB, 98 batches), Output: 100000 rows (2.29MB, 0 batches), Cpu time: 619.00us, Blocked wall time: 2.62ms, Peak memory: 0B, Memory allocations: 0, Threads: 1
              queuedWallNanos    sum: 162.00us, count: 2, min: 53.00us, max: 109.00us
           LocalExchange: Input: 100000 rows (2.29MB, 0 batches), Output: 100000 rows (2.29MB, 98 batches), Cpu time: 326.00us, Blocked wall time: 7.65s, Peak memory: 0B, Memory allocations: 0, Threads: 1
              queuedWallNanos    sum: 28.00us, count: 1, min: 28.00us, max: 28.00us
          -- Aggregation[PARTIAL [supplier_no] total_revenue := sum(ROW["part_revenue"])] -> supplier_no:BIGINT, total_revenue:DOUBLE
             Output: 100000 rows (2.29MB, 98 batches), Cpu time: 767.41ms, Blocked wall time: 0ns, Peak memory: 6.00MB, Memory allocations: 336, Threads: 1
            -- Project[expressions: (supplier_no:BIGINT, ROW["l_suppkey"]), (part_revenue:DOUBLE, multiply(ROW["l_extendedprice"],minus(1,ROW["l_discount"])))] -> supplier_no:BIGINT, part_revenue:DOUBLE
               Output: 2265714 rows (28.35MB, 6011 batches), Cpu time: 1.13s, Blocked wall time: 0ns, Peak memory: 1.00MB, Memory allocations: 5, Threads: 1
                  dataSourceLazyWallNanos    sum: 2.40s, count: 24044, min: 0ns, max: 895.00us
              -- TableScan[table: lineitem, range filters: [(shipdate, BytesRange: [1996-01-01, 1996-03-31] no nulls)]] -> l_suppkey:BIGINT, l_shipdate:VARCHAR, l_extendedprice:DOUBLE, l_discount:DOUBLE
                 Input: 2265714 rows (749.18MB, 0 batches), Raw Input: 59986052 rows (625.41MB), Output: 2265714 rows (69.91MB, 6011 batches), Cpu time: 5.56s, Blocked wall time: 0ns, Peak memory: 24.00MB, Memory allocations: 12972, Threads: 1, Splits: 80
                    dataSourceWallNanos    sum: 5.60s, count: 6091, min: 5.00us, max: 21.21ms
                    localReadBytes         sum: 0B, count: 1, min: 0B, max: 0B
                    numLocalRead           sum: 0, count: 1, min: 0, max: 0
                    numPrefetch            sum: 0, count: 1, min: 0, max: 0
                    numRamRead             sum: 0, count: 1, min: 0, max: 0
                    numStorageRead         sum: 0, count: 1, min: 0, max: 0
                    prefetchBytes          sum: 0B, count: 1, min: 0B, max: 0B
                    ramReadBytes           sum: 0B, count: 1, min: 0B, max: 0B
                    skippedSplitBytes      sum: 0B, count: 1, min: 0B, max: 0B
                    skippedSplits          sum: 0, count: 1, min: 0, max: 0
                    skippedStrides         sum: 0, count: 1, min: 0, max: 0
                    storageReadBytes       sum: 0B, count: 1, min: 0B, max: 0B
    -- Aggregation[SINGLE max_revenue := max(ROW["total_revenue"])] -> max_revenue:DOUBLE
       Output: 1 rows (32B, 1 batches), Cpu time: 606.00us, Blocked wall time: 0ns, Peak memory: 1.00MB, Memory allocations: 2, Threads: 1
      -- Aggregation[FINAL [l_suppkey] total_revenue := sum("total_revenue")] -> l_suppkey:BIGINT, total_revenue:DOUBLE
         Output: 100000 rows (2.29MB, 98 batches), Cpu time: 7.51ms, Blocked wall time: 0ns, Peak memory: 4.00MB, Memory allocations: 140, Threads: 1
        -- LocalPartition[GATHER] -> l_suppkey:BIGINT, total_revenue:DOUBLE
           Output: 200000 rows (4.59MB, 98 batches), Cpu time: 1.01ms, Blocked wall time: 7.65s, Peak memory: 0B, Memory allocations: 0
           LocalPartition: Input: 100000 rows (2.29MB, 98 batches), Output: 100000 rows (2.29MB, 0 batches), Cpu time: 641.00us, Blocked wall time: 2.29ms, Peak memory: 0B, Memory allocations: 0, Threads: 1
              queuedWallNanos    sum: 145.00us, count: 2, min: 55.00us, max: 90.00us
           LocalExchange: Input: 100000 rows (2.29MB, 0 batches), Output: 100000 rows (2.29MB, 98 batches), Cpu time: 374.00us, Blocked wall time: 7.65s, Peak memory: 0B, Memory allocations: 0, Threads: 1
              queuedWallNanos    sum: 47.00us, count: 1, min: 47.00us, max: 47.00us
          -- Aggregation[PARTIAL [l_suppkey] total_revenue := sum(ROW["part_revenue"])] -> l_suppkey:BIGINT, total_revenue:DOUBLE
             Output: 100000 rows (2.29MB, 98 batches), Cpu time: 766.41ms, Blocked wall time: 0ns, Peak memory: 6.00MB, Memory allocations: 336, Threads: 1
            -- Project[expressions: (l_suppkey:BIGINT, ROW["l_suppkey"]), (part_revenue:DOUBLE, multiply(ROW["l_extendedprice"],minus(1,ROW["l_discount"])))] -> l_suppkey:BIGINT, part_revenue:DOUBLE
               Output: 2265714 rows (28.35MB, 6011 batches), Cpu time: 1.13s, Blocked wall time: 0ns, Peak memory: 1.00MB, Memory allocations: 5, Threads: 1
                  dataSourceLazyWallNanos    sum: 2.40s, count: 24044, min: 0ns, max: 687.00us
              -- TableScan[table: lineitem, range filters: [(shipdate, BytesRange: [1996-01-01, 1996-03-31] no nulls)]] -> l_suppkey:BIGINT, l_shipdate:VARCHAR, l_extendedprice:DOUBLE, l_discount:DOUBLE
                 Input: 2265714 rows (749.18MB, 0 batches), Raw Input: 59986052 rows (625.41MB), Output: 2265714 rows (69.91MB, 6011 batches), Cpu time: 5.56s, Blocked wall time: 0ns, Peak memory: 24.00MB, Memory allocations: 12972, Threads: 1, Splits: 80
                    dataSourceWallNanos    sum: 5.60s, count: 6091, min: 5.00us, max: 21.42ms
                    localReadBytes         sum: 0B, count: 1, min: 0B, max: 0B
                    numLocalRead           sum: 0, count: 1, min: 0, max: 0
                    numPrefetch            sum: 0, count: 1, min: 0, max: 0
                    numRamRead             sum: 0, count: 1, min: 0, max: 0
                    numStorageRead         sum: 0, count: 1, min: 0, max: 0
                    prefetchBytes          sum: 0B, count: 1, min: 0B, max: 0B
                    ramReadBytes           sum: 0B, count: 1, min: 0B, max: 0B
                    skippedSplitBytes      sum: 0B, count: 1, min: 0B, max: 0B
                    skippedSplits          sum: 0, count: 1, min: 0, max: 0
                    skippedStrides         sum: 0, count: 1, min: 0, max: 0
                    storageReadBytes       sum: 0B, count: 1, min: 0B, max: 0B

Reply:
The join order is changed now and the plan stats are as follows:
```
Execution time: 7.79s
Splits total: 230, finished: 230
-- OrderBy[s_suppkey ASC NULLS LAST] -> s_suppkey:BIGINT, s_name:VARCHAR, s_address:VARCHAR, s_phone:VARCHAR, total_revenue:DOUBLE
   Output: 1 rows (143.94KB, 1 batches), Cpu time: 1.23ms, Blocked wall time: 0ns, Peak memory: 1.00MB, Memory allocations: 12, Threads: 1
  -- HashJoin[INNER s_suppkey=supplier_no] -> s_suppkey:BIGINT, s_name:VARCHAR, s_address:VARCHAR, s_phone:VARCHAR, total_revenue:DOUBLE
     Output: 1 rows (47.97KB, 1 batches), Cpu time: 137.00us, Blocked wall time: 7.77s, Peak memory: 2.00MB, Memory allocations: 9
     HashBuild: Input: 1 rows (23B, 1 batches), Output: 0 rows (0B, 0 batches), Cpu time: 98.00us, Blocked wall time: 0ns, Peak memory: 1.00MB, Memory allocations: 2, Threads: 1
        distinctKey0       sum: 2, count: 1, min: 2, max: 2
        queuedWallNanos    sum: 75.00us, count: 1, min: 75.00us, max: 75.00us
        rangeKey0          sum: 2, count: 1, min: 2, max: 2
     HashProbe: Input: 1 rows (47.91KB, 1 batches), Output: 1 rows (47.97KB, 1 batches), Cpu time: 39.00us, Blocked wall time: 7.77s, Peak memory: 1.00MB, Memory allocations: 7, Threads: 1
        dataSourceLazyWallNanos    sum: 1.16ms, count: 3, min: 262.00us, max: 509.00us
        dynamicFiltersProduced     sum: 1, count: 1, min: 1, max: 1
        queuedWallNanos            sum: 18.00us, count: 1, min: 18.00us, max: 18.00us
    -- TableScan[table: supplier] -> s_suppkey:BIGINT, s_name:VARCHAR, s_address:VARCHAR, s_phone:VARCHAR
       Input: 1 rows (47.91KB, 0 batches), Raw Input: 53785 rows (47.46MB), Output: 1 rows (47.91KB, 1 batches), Cpu time: 14.25ms, Blocked wall time: 0ns, Peak memory: 2.00MB, Memory allocations: 197, Threads: 1, Splits: 70
          dataSourceWallNanos       sum: 6.92ms, count: 76, min: 2.00us, max: 1.58ms
          dynamicFiltersAccepted    sum: 1, count: 1, min: 1, max: 1
          localReadBytes            sum: 0B, count: 1, min: 0B, max: 0B
          numLocalRead              sum: 0, count: 1, min: 0, max: 0
          numPrefetch               sum: 0, count: 1, min: 0, max: 0
          numRamRead                sum: 0, count: 1, min: 0, max: 0
          numStorageRead            sum: 0, count: 1, min: 0, max: 0
          prefetchBytes             sum: 0B, count: 1, min: 0B, max: 0B
          ramReadBytes              sum: 0B, count: 1, min: 0B, max: 0B
          skippedSplitBytes         sum: 912.76KB, count: 1, min: 912.76KB, max: 912.76KB
          skippedSplits             sum: 30, count: 1, min: 30, max: 30
          skippedStrides            sum: 3, count: 1, min: 3, max: 3
          storageReadBytes          sum: 0B, count: 1, min: 0B, max: 0B
    -- HashJoin[INNER total_revenue=max_revenue] -> supplier_no:BIGINT, total_revenue:DOUBLE
       Output: 1 rows (23B, 1 batches), Cpu time: 2.27ms, Blocked wall time: 7.75s, Peak memory: 2.00MB, Memory allocations: 5
       HashBuild: Input: 1 rows (32B, 1 batches), Output: 0 rows (0B, 0 batches), Cpu time: 76.00us, Blocked wall time: 0ns, Peak memory: 1.00MB, Memory allocations: 2, Threads: 1
          distinctKey0       sum: 1, count: 1, min: 1, max: 1
          queuedWallNanos    sum: 68.00us, count: 1, min: 68.00us, max: 68.00us
       HashProbe: Input: 100000 rows (2.29MB, 98 batches), Output: 1 rows (23B, 1 batches), Cpu time: 2.20ms, Blocked wall time: 7.75s, Peak memory: 1.00MB, Memory allocations: 3, Threads: 1
          queuedWallNanos    sum: 20.00us, count: 1, min: 20.00us, max: 20.00us
      -- Aggregation[FINAL [supplier_no] total_revenue := sum("total_revenue")] -> supplier_no:BIGINT, total_revenue:DOUBLE
         Output: 100000 rows (2.29MB, 98 batches), Cpu time: 6.32ms, Blocked wall time: 0ns, Peak memory: 4.00MB, Memory allocations: 140, Threads: 1
        -- LocalPartition[GATHER] -> supplier_no:BIGINT, total_revenue:DOUBLE
           Output: 200000 rows (4.59MB, 98 batches), Cpu time: 826.00us, Blocked wall time: 12.93ms, Peak memory: 0B, Memory allocations: 0
           LocalPartition: Input: 100000 rows (2.29MB, 98 batches), Output: 100000 rows (2.29MB, 0 batches), Cpu time: 501.00us, Blocked wall time: 12.93ms, Peak memory: 0B, Memory allocations: 0, Threads: 1
              queuedWallNanos    sum: 121.00us, count: 2, min: 58.00us, max: 63.00us
           LocalExchange: Input: 100000 rows (2.29MB, 0 batches), Output: 100000 rows (2.29MB, 98 batches), Cpu time: 325.00us, Blocked wall time: 0ns, Peak memory: 0B, Memory allocations: 0, Threads: 1
          -- Aggregation[PARTIAL [supplier_no] total_revenue := sum(ROW["part_revenue"])] -> supplier_no:BIGINT, total_revenue:DOUBLE
             Output: 100000 rows (2.29MB, 98 batches), Cpu time: 745.80ms, Blocked wall time: 0ns, Peak memory: 7.00MB, Memory allocations: 336, Threads: 1
            -- Project[expressions: (supplier_no:BIGINT, ROW["l_suppkey"]), (part_revenue:DOUBLE, multiply(ROW["l_extendedprice"],minus(1,ROW["l_discount"])))] -> supplier_no:BIGINT, part_revenue:DOUBLE
               Output: 2265714 rows (28.35MB, 6011 batches), Cpu time: 1.11s, Blocked wall time: 0ns, Peak memory: 1.00MB, Memory allocations: 5, Threads: 1
                  dataSourceLazyWallNanos    sum: 2.35s, count: 24044, min: 0ns, max: 545.00us
              -- TableScan[table: lineitem, range filters: [(shipdate, BytesRange: [1996-01-01, 1996-03-31] no nulls)]] -> l_suppkey:BIGINT, l_shipdate:VARCHAR, l_extendedprice:DOUBLE, l_discount:DOUBLE
                 Input: 2265714 rows (749.18MB, 0 batches), Raw Input: 59986052 rows (625.41MB), Output: 2265714 rows (69.91MB, 6011 batches), Cpu time: 5.63s, Blocked wall time: 0ns, Peak memory: 24.00MB, Memory allocations: 12972, Threads: 1, Splits: 80
                    dataSourceWallNanos    sum: 5.75s, count: 6091, min: 5.00us, max: 16.00ms
                    localReadBytes         sum: 0B, count: 1, min: 0B, max: 0B
                    numLocalRead           sum: 0, count: 1, min: 0, max: 0
                    numPrefetch            sum: 0, count: 1, min: 0, max: 0
                    numRamRead             sum: 0, count: 1, min: 0, max: 0
                    numStorageRead         sum: 0, count: 1, min: 0, max: 0
                    prefetchBytes          sum: 0B, count: 1, min: 0B, max: 0B
                    ramReadBytes           sum: 0B, count: 1, min: 0B, max: 0B
                    skippedSplitBytes      sum: 0B, count: 1, min: 0B, max: 0B
                    skippedSplits          sum: 0, count: 1, min: 0, max: 0
                    skippedStrides         sum: 0, count: 1, min: 0, max: 0
                    storageReadBytes       sum: 0B, count: 1, min: 0B, max: 0B
      -- Aggregation[SINGLE max_revenue := max(ROW["total_revenue"])] -> max_revenue:DOUBLE
         Output: 1 rows (32B, 1 batches), Cpu time: 603.00us, Blocked wall time: 0ns, Peak memory: 1.00MB, Memory allocations: 2, Threads: 1
        -- Aggregation[FINAL [l_suppkey] total_revenue := sum("total_revenue")] -> l_suppkey:BIGINT, total_revenue:DOUBLE
           Output: 100000 rows (2.29MB, 98 batches), Cpu time: 6.19ms, Blocked wall time: 0ns, Peak memory: 4.00MB, Memory allocations: 140, Threads: 1
          -- LocalPartition[GATHER] -> l_suppkey:BIGINT, total_revenue:DOUBLE
             Output: 200000 rows (4.59MB, 98 batches), Cpu time: 853.00us, Blocked wall time: 7.75s, Peak memory: 0B, Memory allocations: 0
             LocalPartition: Input: 100000 rows (2.29MB, 98 batches), Output: 100000 rows (2.29MB, 0 batches), Cpu time: 574.00us, Blocked wall time: 2.35ms, Peak memory: 0B, Memory allocations: 0, Threads: 1
                queuedWallNanos    sum: 88.00us, count: 2, min: 26.00us, max: 62.00us
             LocalExchange: Input: 100000 rows (2.29MB, 0 batches), Output: 100000 rows (2.29MB, 98 batches), Cpu time: 279.00us, Blocked wall time: 7.75s, Peak memory: 0B, Memory allocations: 0, Threads: 1
                queuedWallNanos    sum: 62.00us, count: 1, min: 62.00us, max: 62.00us
            -- Aggregation[PARTIAL [l_suppkey] total_revenue := sum(ROW["part_revenue"])] -> l_suppkey:BIGINT, total_revenue:DOUBLE
               Output: 100000 rows (2.29MB, 98 batches), Cpu time: 745.33ms, Blocked wall time: 0ns, Peak memory: 6.00MB, Memory allocations: 336, Threads: 1
              -- Project[expressions: (l_suppkey:BIGINT, ROW["l_suppkey"]), (part_revenue:DOUBLE, multiply(ROW["l_extendedprice"],minus(1,ROW["l_discount"])))] -> l_suppkey:BIGINT, part_revenue:DOUBLE
                 Output: 2265714 rows (28.35MB, 6011 batches), Cpu time: 1.12s, Blocked wall time: 0ns, Peak memory: 1.00MB, Memory allocations: 5, Threads: 1
                    dataSourceLazyWallNanos    sum: 2.35s, count: 24044, min: 0ns, max: 585.00us
                -- TableScan[table: lineitem, range filters: [(shipdate, BytesRange: [1996-01-01, 1996-03-31] no nulls)]] -> l_suppkey:BIGINT, l_shipdate:VARCHAR, l_extendedprice:DOUBLE, l_discount:DOUBLE
                   Input: 2265714 rows (749.18MB, 0 batches), Raw Input: 59986052 rows (625.41MB), Output: 2265714 rows (69.91MB, 6011 batches), Cpu time: 5.64s, Blocked wall time: 0ns, Peak memory: 24.00MB, Memory allocations: 12972, Threads: 1, Splits: 80
                      dataSourceWallNanos    sum: 5.74s, count: 6091, min: 5.00us, max: 15.97ms
                      localReadBytes         sum: 0B, count: 1, min: 0B, max: 0B
                      numLocalRead           sum: 0, count: 1, min: 0, max: 0
                      numPrefetch            sum: 0, count: 1, min: 0, max: 0
                      numRamRead             sum: 0, count: 1, min: 0, max: 0
                      numStorageRead         sum: 0, count: 1, min: 0, max: 0
                      prefetchBytes          sum: 0B, count: 1, min: 0B, max: 0B
                      ramReadBytes           sum: 0B, count: 1, min: 0B, max: 0B
                      skippedSplitBytes      sum: 0B, count: 1, min: 0B, max: 0B
                      skippedSplits          sum: 0, count: 1, min: 0, max: 0
                      skippedStrides         sum: 0, count: 1, min: 0, max: 0
                      storageReadBytes       sum: 0B, count: 1, min: 0B, max: 0B
```

Latest numbers are updated in the PR description

Reply:
@pramodsatya My reading of the plans you shared is that the original execution time is 7.71s and the new one is 7.79s.  Looks like the new time is longer. Is this expected? Would you update PR description to include the plan with stats?

Reply:
The plan stats are printed only for the first run and the plan change doesn't seem to offer any benefits looking at this only. When the time taken is averaged over 5 runs, the avg time reduces from 7.96s to 7.54s. Subsequent runs with the modified query plan take lesser time as seen:
Iteration:  1 => 7.79
Iteration:  2 => 7.43
Iteration:  3 => 7.45
Iteration:  4 => 7.5
Iteration:  5 => 7.55

Thanks for the feedback, will include the plan analysis in the PR description going forward.